enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,188 @@
+
+A new `apiServer` level section will be added to the configuration called `namedCertificates`:
+
+```yaml
+apiServer:
+  namedCertificates:
+  - certPath: ~/certs/api_fqdn_1.crt
+    keyPath:  ~/certs/api_fqdn_1.key
+  - certPath: ~/certs/api_fqdn_2.crt
+    keyPath:  ~/certs/api_fqdn_2.key
+    names:
+    - api_fqdn_1     
+    - *.apps.external.com
+
+```  
+For each provided certificate, the following configuration is proposed:
+1. `names` - optional list of explicit DNS names (leading wildcards allowed).
+   If no names are provided, the implicit names will be extracted from the certificates.
+1. `certPath` -  certificate full path.
+1. `keyPath` -   certificate key full path.
+
+Certificate files will be read from their configured location by the Microshift service,
+each certification file will be validated (see validation rules).
+
+Those certificate will extend the default  API server [external](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L194) certificate configuration.
+
+
+### Workflow Description
+
+#### Default API Certs
+1. by default when there is no namedCertificates configuration the behaviour will remain the same.
+
+#### when custom namedCertificates configured
+1. Device Administrator copies the certificated to MicroShift host
+1. Device Administrator configures `namedCertificates` in the Microshift configuration yaml file (/etc/microshift/config.yaml).
+1. Device Administrator start/restarts MicroShift
+1. During startup the certifates paths will be checked and validated. in case of an error service will produce clear error log and exit.
+1. Certifates will be passed to the tls-sni-cert-key command line flag preceding all the other certificates.
+1. kube-apiserver picks up the certificates and start serving them on the configured SNI.
+
+### Implementation Details
+
+The certs will be prepended to the []configv1.NamedCertificate list before the api server is started. (it will be added to `-tls-sni-cert-key` flag)
+
+this certification paths configuration and names will be prepended into the kube-apiserver `tls-sni-cert-key` command line flag.
+when same SNI appear in the CN part of the provided certs,this certificate will take precedence over the `default` external-signer. [ref](https://github.com/kubernetes/kubernetes/blob/98358b8ce11b0c1878ae7aa1482668cb7a0b0e23/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go#L38)
+
+
+### API Extensions
+N/A
+
+### Risks and Mitigations
+* User provide incorrect/expired certificate.
+  kube-api server will continue serving with an expired cert - similiar approach is taken by OpenShift.
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
+
+* User provided expired Certificate and Microshift service started/restarted 
+  Microshift will start with a warning in the logs,kube-api will continue serve with an expired cert - similiar approach is taken by OpenShift.
```

Comment:
I think that we should have an option for security conscious users which would configure microshift to not start when the certs provided are invalid (e.g. when they've been revoked, expired, etc.).

Reply:
so will adding flag like `allow_invalid_certificate`  globally for all the NameCerts will work  ?

Reply:
We'll want it to be an enum, rather than a boolean. What are the options? Something like:

* Run using expired certs
* Run but do not use expired certs (so only the self-signed certs are used)
* Refuse to run

Are there other behaviors?

What should be the default?

Reply:
custom-certificates only affect on how users access the  API  server externally (using oc client),so it shouldn't affect any internal functionality nor  cause any data loss.

Reply:
@dhellmann , @jerpeter1  i've added the new configuration option `InvalidCertificatesPolicy` , PTAL

Reply:
We'll want more text explaining what the valid values of that new configuration setting are, how it maps to the OpenShift API for doing validation, etc.

Reply:
I'll resolve this thread, since the text it's discussing has been removed.